File: src/Starter/Features/Configuration/SettingsWindow/SettingsWindowViewModel.fs
namespace Starter.Features.Config.UI.SettingsWindow

open System
open System.Collections.Generic
open Starter.Features
open Starter.SearchEngine
open Starter.Features.Config

open Avalonia.Controls
open FluentAvalonia.UI.Controls
open FluentIcons.Common
open FluentIcons.Avalonia.Fluent

open ReactiveUI
open R3

type MenuItemVM =
    { Id: string
      Icon: IconSource | null
      Name: string
      Control: Control }

    static member create control (se: SearchEngine) =
        { Id = se.Id
          Icon = se.Icon |> StarterIconSource.buildIconSource
          Name = se.Name
          Control = control }

type WindowViewModel(baseConfig, searchEngines: Dictionary<string, SearchEngine> BehaviorSubject) =
    inherit ReactiveObject()

    let starterSettingsVM = new UI.StarterSettings.ViewModel(baseConfig, searchEngines)
    let starterSettingsMenuItem =
        { Id = "starter-settings"
          Icon = FluentIconSource(Icon = Icon.Settings, IconSize = IconSize.Size16)
          Name = "Starter settings"
          Control = UI.StarterSettings.StarterSettings(DataContext = starterSettingsVM) }

    let logsVM = LoggingView.LogsViewModel()
    let logsMenuItem =
        { Id = "starter-logs"
          Icon = FluentIconSource(Icon = Icon.DocumentText, IconSize = IconSize.Size16)
          Name = "Logs"
          Control = LoggingView.LogsView(DataContext = logsVM) }

    let mutable selectedPage = starterSettingsMenuItem
    let menuItems = new BehaviorSubject<_ array>(Array.empty)

    let sub =
        searchEngines.ObserveOnUIThreadDispatcher()
        |> Observable.subscribe (fun d ->
            d
            |> Seq.choose (fun kv ->
                menuItems.Value
                |> Array.tryFind (fun i -> i.Id = kv.Key)
                |> Option.map Some
                |> Option.defaultWith (fun () ->
                    match kv.Value.LoadSettingsControl() with
                    | null -> None
                    | control -> kv.Value |> MenuItemVM.create control |> Some
                )
            )
            |> Seq.sortBy _.Name
            |> Seq.append [ starterSettingsMenuItem; logsMenuItem ]
            |> Seq.toArray
            |> menuItems.OnNext
        )

    interface IDisposable with
        member _.Dispose() = sub.Dispose()

    member this.Configuration = starterSettingsVM.Configuration

    member this.MenuItems = menuItems
    member this.SelectedPage
        with get () = selectedPage
        and set v = this.RaiseAndSetIfChanged(&selectedPage, v) |> ignore

    member this.SelectSettingsPage() = this.SelectedPage <- starterSettingsMenuItem
    member this.SelectLogsPage() = this.SelectedPage <- logsMenuItem

File: src/Starter.Desktop/ViewModels/SettingsWindowViewModel.cs
using Avalonia.Controls;
using R3;
using Starter.Features.Config;
using Starter.SearchEngine;

namespace Starter.Desktop.ViewModels;

public partial class MenuItemViewModel(
    string title,
    StarterIconSource icon,
    Control control
) : ObservableObject
{
    public string Title { get; init; } = title;
    public Control Control { get; init; } = control;
    [ObservableProperty] private StarterIconSource icon = icon;

    public MenuItemViewModel(SearchEngine.SearchEngine engine, Control control) : this(engine.Name, engine.Icon, control) =>
        engine.Changed += (_, _) => Icon = engine.Icon;
}

public partial class SettingsWindowViewModel : ObservableObject
{
    [ObservableProperty] private List<MenuItemViewModel> pages = [];
    [ObservableProperty] private MenuItemViewModel selectedPage; // Currently selected settings page

    private readonly SettingsViewModel settingsVm;
    public BehaviorSubject<Configuration> Config => settingsVm.Config;
    public IObservable<Configuration> ConfigSystemObservable { get; private set; }

    public SettingsWindowViewModel(Configuration config, SearchEngineStore searchEngineStore)
    {
        settingsVm = new SettingsViewModel(config, searchEngineStore);
        ConfigSystemObservable = Config.AsSystemObservable();

        // Add Starter settings
        Pages.Add(new MenuItemViewModel(
            "Starter settings",
            Icons.Settings,
            new Views.Settings { DataContext = settingsVm }
        ));
        Pages.Add(new MenuItemViewModel(
            "Logs",
            Icons.Logs,
            new Views.Logs { DataContext = new LogsViewModel() }
        ));

        // Set default page
        selectedPage = pages[0]; // starter settings

        // Add search engine settings
        foreach (var kv in searchEngineStore.SearchEngines)
        {
            var control = kv.Value.LoadSettingsControl();
            if (control == null) continue;
            Pages.Add(new MenuItemViewModel(kv.Value, control));
        }
    }
}
